### ISSUE TITLE
Proposal to hide implementation of `NgIf`, `NgFor`

### ISSUE THREAD

**Issue by matanlurey** [dart-lang/angular#1557]
Already we have, mostly undocumented, [compiler intrinsics][1] around `NgIf` in some cases:

https://github.com/dart-lang/angular/blob/67d1bd06a152956e315d443042ef9da646e57552/angular/lib/src/compiler/view_compiler/property_binder.dart#L465-L481

https://github.com/dart-lang/angular/blob/aea9daa80c7d9bb87b824cc3dd0d79944c2144cd/angular/lib/src/compiler/view_compiler/compile_view.dart#L281-L287

I'd like to propose to slowly make `NgIf` and `NgFor` compiler-implemented classes, i.e.:

```dart
abstract class NgIf {
  external set ngIf(bool ngIf);
}
```

In the short term we could have a hidden concrete implementation:

```dart
class InternalNgIf implements NgIf { ... }
```

... and eventually we would just emit different code for `*ngIf` all-together. It would be more complicated (and breaking) but we could do a similar path for `*ngFor`. We might even want to consider just renaming these `@if` and `@for` and making them built-ins instead of structural directives.

/cc @leonsenft @alorenzen 

[1]: https://en.wikipedia.org/wiki/Intrinsic_function

**Comment by leonsenft** [comment]
Yeah, we also have some NgFor specific optimizations built into the compiler but they're pretty limited. Not being restricted by the fact it's a directive could potentially make the optimizations better, and easier to implement.
